### ISSUE TITLE
AutoComplete: Custom rendering with links causes CSP issues

### ISSUE THREAD

**Issue by martin654** [primefaces/primefaces#10654]
### Describe the bug

Hi,
we have autocomplete component with commandLinks used as autocomplete result items:
![obrazek](https://github.com/primefaces/primefaces/assets/22746146/56b6521f-664b-4ea1-86db-4bff31d22d21)

Without CSP it works fine. When we turn CSP on, it reports problems in JS console.

First errors occur when text entered in autocompelte and reuslts are rendered in page (each result item has it's error)
Second type of error occurs when item clicked.

![obrazek](https://github.com/primefaces/primefaces/assets/22746146/f4041680-5dbb-4313-938f-deeaa8c1e2e2)

Please see reproducer project attached below.

### Reproducer

XHTML snippet:
```
<h:form>
            <p:commandButton id="leftmenuSearchButton" type="button" icon="fa fa-search" />

            <p:overlayPanel for="leftmenuSearchButton" widgetVar="leftMenuSearchPanel" at="right top"
                            onHide="PF('leftMenuSearchAutocompleteWidgetVar').clear();">
                <p:autoComplete id="leftMenuSearchAutocomplete" var="_leftMenuSearchResultItem"
                                widgetVar="leftMenuSearchAutocompleteWidgetVar" minQueryLength="1"
                                completeMethod="#{testView.searchAutocomplete}">
                    <p:ajax event="query" ignoreAutoUpdate="true"/>
                    <p:column>
                        <p:commandLink action="#{testView.invokeCommand}" oncomplete="PF('leftMenuSearchPanel').hide();">
                            <h:outputText value="#{_leftMenuSearchResultItem.label}" />
                            <f:param name="commandLinkItemId" value="#{_leftMenuSearchResultItem.id}"/>

                            <p:confirm source="this" header="Really?" message="Are you sure?"
                                       disabled="#{not _leftMenuSearchResultItem.showConfirm}"
                                       icon="ui-icon-alert"/>
                        </p:commandLink>
                    </p:column>
                </p:autoComplete>
            </p:overlayPanel>
        </h:form>
```

Whole reproducer project:
[pf-xxxx.zip](https://github.com/primefaces/primefaces/files/12587705/pf-xxxx.zip)

### Expected behavior

I would expect that it would work with or without CSP the same.

### PrimeFaces edition

Elite

### PrimeFaces version

13.0.1

### JSF implementation

Mojarra

### JSF version

2.3.17

### Java version

11

**Comment by melloware** [comment]
@martin654 is this a new issue or was it this way with 12.0.0?  Or did you just discover it today?

**Comment by melloware** [comment]
I use CSP on all my apps in PRD over 70+ apps and counting...

let me see if I can figure this one out but I have a feeling it might be p:confirm related.

**Comment by melloware** [comment]
We also run our Full 700+ integration tests with CSP both on an off every night here at PrimeFaces. See: https://github.com/primefaces/primefaces/actions/runs/6155556455

**Comment by melloware** [comment]
OK the issue with CSP is because of this fix: https://github.com/primefaces/primefaces/issues/7763

Because its just using `FastStringWriter` its not going through CSPWriter and thus using an `onClick` script which is not allowed.

Similar to this issue: https://github.com/primefaces-extensions/primefaces-extensions/issues/964

We would need a way of FastStringWriter to do the same logic the CSP Writer does or somehow wrap it?

**Comment by melloware** [comment]
OK so right now this won't work with CSP because of the custom rendering and the fact you aren't using it just for display but you have an action when you click on an item.  Your link is what is exposing this issue.

The only way to TRULY fix this for CSP is that this "custom" rendering feature would need to be rewritten to use UIData or some repeating element to repeat properly in JSF.

IT is from this issue: https://github.com/primefaces/primefaces/issues/8765 which we fixed the ID's.

But it still would have caused a CSP problem as you saw in 12.0.0 because all of those ID's for all elements are the same.  So CSP can only register the on Click on the first element in the list since each element is NOT uniquely identified.

Thomas originally said 

> this should be done with UIRepeat/NamingContainer 

My response was:

> I honestly looked all over for examples and studied the code and I could not figure out how to make these components easily support that for this "Advanced" rendering feature. Any hints would be great but I think we might need to go with this short term solution until a longer term solution can be found?

**Comment by melloware** [comment]
Also @martin654 what is your use case for an AutoComplete with links?  It seems like an odd Ui/UX paradigm?

**Comment by martin654** [comment]
@melloware our use case is to provide quick access for users to almost all functions (screen flows) in our app. So user types for example "prod" and as a result we provide links to "product overview" and "new product".

More detailed explanation:
First, as a left menu we use MegaMenu component. But it is used as "most common functions" or "global favorites" items with about 70 items in it.  Second thing is a separate screen with a searchable tree component with all possible features accessible from our app visualized hiearchically (about 300 items). But this is "far away" for users which know exactly what they are searching for. So for those experienced user we have a button above our left menu with overlay in which we use the autocomplete component to provide access to all functions using text search.

It was a users request and they happily use it for about 2 years in PROD environment.

**Comment by melloware** [comment]
OK its just strange you are using an Input component like AutoComplete NOT to solicit input but for user navigation?

You want something more like the PrimeFaces showcase has to find an item to select it for navigation right?

![image](https://github.com/primefaces/primefaces/assets/4399574/d85b70fc-b378-43a6-822f-8c62f394c31c)

**Comment by martin654** [comment]
Yep. Showcase looks like autocomplete too. 

I need commandLink component for this because some of them are set with ajax=false and some of them with ajax=true depending on whether user stays in same window or new window /tab is opened. We use spring webflow under the hood and this type of navigation is performed with ajax=false.

**Comment by martin654** [comment]
@melloware  that is great news.

I can test snapshot in couple of days if needed. Thanks

**Comment by melloware** [comment]
@martin654 you can test your reproducer with 14.0.0-SNAPSHOT I published to maven central.  Just add this to your pom.xml

```xml
<repositories>
  <repository>
    <id>sonatype-snapshots</id>
    <name>Sonatype Snapshot Repository</name>
    <url>https://oss.sonatype.org/content/repositories/snapshots</url>
    <releases>
        <enabled>false</enabled>
    </releases>
    <snapshots>
        <enabled>true</enabled>
    </snapshots>
  </repository>
</repositories>
```

**Comment by martin654** [comment]
@melloware 
Hi, I tested reproducer provided above with 14.0.0-SNAPSHOT with these results:
- without CSP it works as expected
- with CSP second item (without confirm) doesn't invoke invokeCommand method

I tested it also in our production app with same results. Confirm makes it work (action is invoked) somehow.

**Comment by melloware** [comment]
Your second item is the disabled one right?   I thought we confirmed that was an issue with 12.0.0 without CSP so I think it is a different issue?

**Comment by martin654** [comment]
Second item is not disabled, it is just without confirm (confirm rendered attribute evaluated as false for second item). 

I think it is still CSP related, because there is a difference between CSP on and off. And yes, your fix made it much better and this seems like a little bit different issue.

**Comment by melloware** [comment]
Ok let me see if I can debug but it strange that it's related to confirm being rendered or not.

**Comment by melloware** [comment]
OK I was able to fix it.

**Comment by melloware** [comment]
OK I just pushed 14.0.0-SNAPSHOT to Central if you want to `mvn -U` and try your code again?

**Comment by martin654** [comment]
@melloware  I tested the snapshot with the reproducer and with our PROD app both with and without CSP. Works like a charm! Thank you for the fix! 

I am looking forward to 13.0.2 version.

Best regards

**Comment by melloware** [comment]
OK I am merging this fix into 13.0.2!
